lib/grandpa/message_tracker.go

```diff
@@ -122,3 +127,32 @@ func (t *tracker) handleBlock(b *types.Block) {
 		delete(t.commitMessages, h)
 	}
 }
+
+func (t *tracker) handleTick() {
+	t.mapLock.Lock()
+	defer t.mapLock.Unlock()
+
+	for _, vms := range t.voteMessages {
+		for _, v := range vms {
+			// handleMessage would never error for vote message
+			_, err := t.handler.handleMessage(v.from, v.msg)
+			if err != nil {
+				logger.Debugf("failed to handle vote message %v: %s", v, err)
+			}
+
+			if v.msg.Round < t.handler.grandpa.state.round && v.msg.SetID == t.handler.grandpa.state.setID {
+				delete(t.voteMessages, v.msg.Message.Hash)
```

Comment:
I think it is not a good idea to modify the map that is being iterated, we can create a slice with hashes to delete and remove them right after the iteration finishes